modules.md

```diff
@@ -0,0 +1,233 @@
+
+Note that in the future, types will also be in a separate namespace.
+
+
+## Importing from other directories within the project
+
+File paths (as well as filenames) may be used inside the quoted string of an `import`. Paths may contain `/` (but not `\`, even on Windows). Paths are always relative to the top-level of the project (not relative to the current file or absolute paths), and may not include `.` or `..` (alternative: we could relax this requirement and allow relative paths as long as traversing the path never leaves the project directory) or global identifiers (`c:`, etc. or starting with '/'). Paths are ignored in the name of the module as imported. E.g., `import "foo.kcl"` brings in `foo` from the root directory of the project, `import "baz/bar/foo.kcl"` brings in `foo` from `foo.kcl` found at `baz/bar/`, `import a from "baz/bar/foo.kcl"` brings in `a` from the same file. The path of the file makes no difference to the imported name or to privacy.
+
+Directories cannot be imported or imported from, e.g., `import foo from "baz/bar"` is not allowed.
+
+All modules within a project are nameable, there is no need for `export` anywhere on a module. If all items in a module are private (i.e., none are `export`ed), the module itself can still be named, but it is useless (like an empty object) and a glob import would import nothing.
+
+## Imports from the standard library
+
+Imports from the standard library work like other imports, but the import path begins with `std` and has no `.kcl` extension. The interpreter knows where to find the standard library which works like any other KCL project.
+
+There is an implicit import `import * from 'std'` in every file unless it includes a `@no_std` attribute. Items in `std` can be imported too. There is no explicit prelude, all of std is treated like a prelude.
+
+
+## Importing from other projects
+
+This is mostly postponed for future work. We would use the same syntax and semantics as the rest of the system, however the string in the import would include some indicator that the target is outside the project and how to locate it.
+
+Example with very, very strawman syntax: `import a from "cargo://some-library/foo.kcl"` would import `a` from the file `foo.kcl` in the project `some-library` located in the `cargo` (lol) repository. `import a from "local://../some-library/foo.kcl` would import `a` from the file `foo.kcl` from a sibling directory of the current project called `some-library`. The point of these examples is not the syntax for identifying other projects or the places where an external project might live, just to show how the `import` statement might be extended to support other projects.
+
+Privacy across multiple projects is postponed, but I would like to keep it simple, e.g., just keeping the single `export` keyword to expose names to other projects as well as other modules in the same project.
```

Comment:
In most programming languages, it's really useful to distinguish between exposed to the project and exposed outside the project, as in `pub` vs. `pub(crate)` in Rust. The most common use case is that you want to expose something from a module for testing, but you don't want to expose it to the world. But it can also come up for helper or util functions for libraries. I think that the lack of this will lead to more things being exported than should, i.e. people learning to default to `export` and private library internals exposed to the world because there's no other way to reuse code.

Reply:
Yeah, I think there is some work to do still for cross-project import/export. (This comment is from quite a while ago, I still think simple is important, but I'm not sure if just `export` is enough). I think we would want something for exporting from a project, but that depends a bit on importing - I'm not sold on the simple strawman of importing a file from a repo - I think we might want some project scoping which would include what is exported.